### PR TITLE
Feature/better find package

### PR DESCRIPTION
**Pull request by victor-uavia**
This PR is a proposal to fix https://github.com/google/or-tools/issues/732 using `find_package` in order to have better granularity for handling dependencies (this is inspired by https://gist.github.com/Som1Lse/6458d28c4bb09163a5389f24fec981db).
Thanks to the built-in variable `<package_name>_DIR` variable we can override the default path for `find_package` in order to use : 
- some preinstall cmake package
- custom source repo (either download with `BUILD_DEPS=ON` or already download and provide by the user through a custom `<Package_name>Config.cmake`)

Allowing to use `find_package` in order to resolve out of tree dependencies allow or-tools to integrate nicely with Yocto, which is not the case today as Yocto since we cannot relay on the `BUILD_DEPS` switch in Yocto (it does not allow to download dependency during cmake configure time).

⚠️ Note that this PR will properly for work preinstall cmake package once https://github.com/abseil/abseil-cpp/pull/182 will be merged

### CODE REVIEW

**Comment by victor-uavia** [comment]
@Mizux do you think this PR could be integrate into the code ?

**Comment by Mizux** [comment]
First, thanks for your contribution !
 
If I understand correctly, you *just* split the `cmake/external/CMakeLists.txt` into several `DependencyConfig.cmake` so for each deps you could choose the system package or ortools will compile it for you.
-> I'm ok with this.

*But* in this case the new macro `_find_package()` seems strange to me 
I mean you `include()` them if needed, while for me you should have modified the `CMAKE_PREFIX_PATH` instead so `find_package()` would find them...

Also, one other side topic: I think I'll have to change the CMake behavior, I would like to install deps first then consume the install, than trying to use it as `add_subdirectory()`. Currently it's conflict with dependency install rule or you need to hack dependencies to make it work...
see: https://travis-ci.org/google/or-tools/jobs/482350303#L904

Sorry for the delay, I'm still thinking about it and where I would like to go.
Need to play with your branch...

File: CMakeLists.txt [hunk 1]
```diff
@@ -79,25 +70,25 @@ message(STATUS "Build .Net Binding: ${BUILD_DOTNET}")
...
-  set(ZLIB_DIR ${CMAKE_CURRENT_SOURCE_DIR}/cmake/external CACHE PATH "ZLIB dependency path")
+  list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake/external/ZLIB)
 endif()
 if(NOT Protobuf_DIR AND BUILD_DEPS)
-  set(Protobuf_DIR ${CMAKE_CURRENT_SOURCE_DIR}/cmake/external CACHE PATH "Protobuf dependency path")
+  list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake/external/Protobuf)
 endif()
 if(NOT abseil_DIR AND BUILD_DEPS)
-  set(abseil_DIR ${CMAKE_CURRENT_SOURCE_DIR}/cmake/external CACHE PATH "abseil dependency path")
+  list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake/external/abseil)
```
Comment (Mizux) [inline CMakeLists.txt]: abseil-cpp ? (I can also do it in a later commit) 
just becaue there is also https://github.com/abseil/abseil-py and maybe we will need it one day too

**Comment by victor-uavia** [comment]
@Mizux I rework the way this branch handle its dependencies in https://github.com/google/or-tools/pull/1019/commits/2a003e0938c15c885ee9976835d22568bbc8534a the branch now donwload, build & *install its file* in the build directory (no more subproject).

I have however the following compilation error which I do not understand and I would like your help to see what I could have missed.

```
/or-tools/or-tools/ortools/data/jobshop_scheduling_parser.cc:26:1: error: C++ requires a type specifier for all declarations
DEFINE_int64(jssp_scaling_up_factor, 100000L,
^
/or-tools/or-tools/ortools/data/jobshop_scheduling_parser.cc:26:14: error: use of undeclared identifier 'jssp_scaling_up_factor'
DEFINE_int64(jssp_scaling_up_factor, 100000L,
             ^
/or-tools/or-tools/ortools/data/jobshop_scheduling_parser.cc:359:45: error: use of undeclared identifier 'FLAGS_jssp_scaling_up_factor'
          static_cast<int64>(round(weight * FLAGS_jssp_scaling_up_factor));
                                            ^
/or-tools/or-tools/ortools/data/jobshop_scheduling_parser.cc:375:19: error: use of undeclared identifier 'FLAGS_jssp_scaling_up_factor'
                  FLAGS_jssp_scaling_up_factor !=
                  ^
/or-tools/or-tools/ortools/data/jobshop_scheduling_parser.cc:385:17: error: use of undeclared identifier 'FLAGS_jssp_scaling_up_factor'
                FLAGS_jssp_scaling_up_factor);
                ^
/or-tools/or-tools/ortools/data/jobshop_scheduling_parser.cc:389:22: error: use of undeclared identifier 'FLAGS_jssp_scaling_up_factor'
              1.0L / FLAGS_jssp_scaling_up_factor);
                     ^
```

**Comment by Mizux** [comment]
Hi,

I revamped your CL in #1116 but with few changes:
* Mostly Keep `BUILD_dependencies` options BUT
  * by default don't build dependencies (easier for distro package maintainer) **unless** you build wrappers
   (ed BTW [PEP 513](https://www.python.org/dev/peps/pep-0513/#the-manylinux1-policy) for python native package doesn't allow us to use system libraries, so if we have to ship them...)
  * you can still enable `BUILD_*` if you want...
  * if BUILD_DEPS=ON all `BUILD_*` options are forced to ON
    https://github.com/google/or-tools/blob/4477092d8d38854b19b577bd586fd2829d0e8838/CMakeLists.txt#L49-L50
   which is the opposite of your
   https://github.com/victor-uavia/or-tools/blob/2a003e0938c15c885ee9976835d22568bbc8534a/CMakeLists.txt#L42-L43

* I don't use a `Find*.cmake` simply call `find_package(* REQUIRED)` and let CMake to [do its stuff](https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure)
  * user can use `CMAKE_PREFIX_PATH`/`CMAKE_MODULE_PATH` if they want to hack etc..

* all deps, if build, are installed in `CMAKE_BINARY_DIR/dependencies/install/` so I just need to append this to `CMAKE_PREFIX_PATH` to see all available prebuilt dependencies.

* I use a template https://github.com/google/or-tools/blob/master/cmake/dependencies/CMakeLists.txt.in for ExternalProject along a function in utils.cmake
  https://github.com/google/or-tools/blob/4477092d8d38854b19b577bd586fd2829d0e8838/cmake/utils.cmake#L66-L126

* I use my new Cbc split repo forks of the official ones

* I use the fixed abseil-cpp name i.e. `absl::base` instead of "broken" `absl::absl_base` since my internal path make its way to the public abseil-cpp repo.

Any feedback are welcome, hope you'll better understand what I'm aiming for....

**Comment by Mizux** [comment]
I'm closing this PR since we have #1116 
Still open to any proposal, critics etc....

**Pull request closed without merging**